Add app handling for elevation profiles stored in projects

nyalldawson opened the pull request:
This continues on work started in https://github.com/qgis/QGIS/pull/62757, and links the app handling of elevation profiles to the project elevation profile manager.

Now, elevation profiles stored in the project can be re-opened after opening a project, and changes made to the profile will be stored in the project.

The elevation profile manager dialog is also fully functional, able to show/rename/remove profiles from the project.

Sponsored by Erftverband

nyalldawson committed:
Rework generation of elevation profile canvas title

nyalldawson committed:
Associate elevation profile dock widget with a profile from the project

nyalldawson committed:
Handle default app settings for new elevation profiles

nyalldawson committed:
Allow existing elevation profiles to be re-opened

nyalldawson committed:
Cleanup widgets when profiles removed

nyalldawson committed:
Correctly restore profile manager geometry

nyalldawson committed:
Fix rename action in elevation profile widget

nyalldawson committed:
Dirty project on profile rename, removal

nyalldawson committed:
Store profile curves in project, and restore

nyalldawson committed:
Store/restore tolerance, axis lock

nyalldawson committed:
Store/restore profile distance units

nyalldawson committed:
Correctly handle stored curve crs, and store crs when updating curve

nyalldawson committed:
Store/restore subsections symbol

nyalldawson committed:
Store/restore elevation profile layers

troopa81 reviewed (commented):
Is there a reason why QgsElevationProfileCanvas doesn't rely on QgsElevationProfile (except that the latter came after the former) ?

Thus, layout could retrieve elevation profile information from the manager and not from the widget/app part, and would avoid the `sBuildCopyMenuFunction` trick.

It's maybe beyond the scope of this PR, but I do think that it would improve the model/view separation.

nyalldawson committed:
Use simpler approach to populating menu

nyalldawson committed:
Dirty project when elevation profile layer set changes

nyalldawson committed:
Const

troopa81 commented:
@nyalldawson as much as I value your experience,  I have learned the hard way that not respecting MVC principle is a source of a lot of bugs (attribute form for instance)

nyalldawson committed:
When initially populating elevation layer tree, don't re-add existing layers

Avoids duplicate layers showing in tree when restoring projects

nyalldawson committed:
Use elevation profile layer tree in dock widget

nyalldawson committed:
[feature] Permit creation of groups in elevation profile layer tree

Allows grouping of layers, with similar logic to main canvas
layer tree groups. Individual groups can be checked or unchecked,
allowing multiple layers to be shown / hidden from the elevation
profile at once.

Just like the main canvas layer tree group handling:

- Groups can be created from a new toolbar action "add group"
- Layers and groups can be rearranged via drag and drop
- Right clicking a group exposes an option to rename the group
- Selecting multiple layers/groups in the view and right click
  exposes and option to "Group Selected"

nyalldawson committed:
New elevation profiles now inherit the main layer tree structure

They now start with the same layer tree structure as the project,
i.e. respecting the initial set of groups

nyalldawson committed:
Add duplicate button to elevation profile manager

Allows duplicating existing profiles. Handy if the user has spent
effort in customising an existing profile (eg rearranging layers
and groups) and wants to re-use that configuration in a new profile.

nyalldawson committed:
Remove redundant nullptr check

nyalldawson commented on src/app/elevation/qgselevationprofilewidget.cpp:
```diff
@@ -413,6 +425,11 @@ QgsElevationProfileWidget::QgsElevationProfileWidget( const QString &name )
+        if ( mProfile )
+        {
+          mProfile->setDistanceUnit( unit );
+          QgsProject::instance()->setDirty();
```
@troopa81 

> I really don't want to nitpick here, but I think this is a matter of importance.

I've been discussing this at length with @nirvn, and I think it's a discussion which belongs elsewhere. This PR isn't the place to decide this, as there's no clear "correct" answer and this PR is doing exactly what a lot of other existing code in master does (dirties the project from app level actions).

I think we should leave as is here, and then start a discussion (and possible follow up QEP policy) on the mailing list to determine the group consensus.

troopa81 replied:
You're right. [Done](https://lists.osgeo.org/pipermail/qgis-developer/2025-November/067870.html)

nyalldawson committed:
Move project dirtying to core

nyalldawson commented:
@alexbruy could i get your eyes on this? :pray:
